ui/index.d.ts

```diff
@@ -360,6 +362,128 @@ export declare abstract class AzureWizardStep<T> {
     public abstract execute(wizardContext: T, outputChannel: OutputChannel): Promise<T>;
 }
 
+export interface ISubscriptionWizardContext {
+    credentials: ServiceClientCredentials;
+    subscription: Subscription;
+}
+
+export interface ILocationWizardContext extends ISubscriptionWizardContext {
+    /**
+     * You may specify the defaultLocationName if you don't want the `LocationStep` to prompt for a location
+     * For example, if the user selects an existing resource, you might want to use that location as the default for the wizard's other resources
+     */
+    defaultLocationName?: string;
+
+    /**
+     * The location to use for new resources
+     * This value will be defined after `LocationStep.prompt` occurs.
+     */
+    location?: Location;
+}
+
+export declare class LocationStep<T extends ILocationWizardContext> extends AzureWizardStep<T> {
+    public prompt(wizardContext: T, ui: IAzureUserInput): Promise<T>;
+    public execute(wizardContext: T, outputChannel: OutputChannel): Promise<T>;
+}
+
+export interface IAzureNamingRules {
+    minLength: number;
+    maxLength: number;
+
+    /**
+     * A RegExp specifying the invalid characters.
+     * For example, /[^a-z0-9]/ would specify that only lowercase, alphanumeric characters are allowed.
+     */
+    invalidCharsRegExp: RegExp;
+
+    /**
+     * Specify this if only lowercase letters are allowed
+     * This is a separate property than `invalidCharsRegExp` because the behavior can be different.
+     * For example, when generating a relatedName, we can convert uppercase letters to lowercase instead of just removing them.
+     */
+    lowercaseOnly?: boolean;
+}
+
+export interface IRelatedNameWizardContext {
+    /**
+     * A task that evaluates to the related name that should be used as the default for other new resources or undefined if a unique name could not be found
+     * The task will be defined after `AzureNameStep.prompt` occurs.
+     */
+    relatedNameTask?: Promise<string | undefined>;
+}
+
+/**
+ * A generic class for a step that specifies the name of a new resource, used to generate a related name for other new resources.
+ * You must implement `isRelatedNameAvailable` and assign `wizardContext.relatedNameTask` to the result of `generateRelatedName`
+ */
+export declare abstract class AzureNameStep<T extends IRelatedNameWizardContext> extends AzureWizardStep<T> {
+    /**
+     * This method will by called by `generateRelatedName` when trying to find a unique suffix for the related name
```

Comment:
What was the difference between isRelatedNameAvailable and isNameAvailable?

Reply:
isNameAvailable is for this resource (for example the naming rules for an App Service when creating an App Service)

isRelatedNameAvailable is used when generating a related name for other resources (aka it would check existing resources groups, app service plans, etc. to see if the generated name is available)